Module 2, Lab 5 - Power 2 - Sample Size Planning Illustration
=============================================================

In this lab, we will practice using the `tt_ind_solve_power` from the Python `statsmodels.stats.power` package for determining the minimum necessary sample size for a two-sample *t*-test design. I illustrate a real-life iterative project planning sequence so you can see how power planning can be integrated into a data-science research project.

First, load the required software:

In [1]:
from statsmodels.stats.power import tt_ind_solve_power

You wish to compare groups at two different locations in your
organization to see if either group is more satisfied with their working
conditions. You will be comparing the groups with a *t*-test, and you
care deeply about estimating the effect, even if it is small. Data will
be challenging to get, however, as you will have to get managers to ask
employees to return surveys. You need to determine the minimum necessary
sample size to get adequate power.

Usually, we start with our dream scenario. Let's ask for 90% power to
detect a very small effect size (*d* = .10). What sample size would be
required?

In [2]:
tt_ind_solve_power(effect_size=0.1, nobs1 = None, alpha=0.05, power=0.9, ratio=1, alternative='two-sided')

2102.44476680053

We see here that we need 2103 people per group, or 4206 people in total.
Knowing the size of the organization, you know anything that size is
likely out of the question. You might be able to get away, you think
with collecting data from 500 participants without imposing too much on
team supervisors. So, you try again, this time with a more realistic 80%
power and a mid-range-small effect of *d* = .25.

In [3]:
tt_ind_solve_power(effect_size=0.25, nobs1 = None, alpha=0.05, power=0.8, ratio=1, alternative='two-sided')

252.1275051543411

By pure happenstance, you get 253 per group (always round up to ensure
sample size is adequate). You take this proposal to collect data from
500 employees to your supervisor; after some discussion, you are told
that they will try to push for a large sample (on the basis of your
request), but they've decided 400 is the maximum they are likely to be
able to collect.

Now, you change your strategy. 400 is close to 500, so it's likely to be
similar. You now leave out effect size and input *n* = 200 (since it's
per group) and a request for 80% power:

In [4]:
tt_ind_solve_power(effect_size=None, nobs1 = 200, alpha=0.05, power=0.8, ratio=1, alternative='two-sided')

0.28083522542893885

You will now have 80% power to detect effects as small as *d* = .28,
which is still a mid-range small effect.

Before you tell everyone that will still work with this suggested sample
size, you run a loop to estimate power at that sample size for various
effect sizes (e.g., power would suffer if *d* were lower, but would it
be *that* terrible if *d* were, say, .10?)

In [5]:
d_values = [x/100.0 for x in range(5,55,5)] # Need range to 55 since Python is zero based indexing

powers = [tt_ind_solve_power(effect_size=d, nobs1 = 200, alpha=0.05, power=None, ratio=1, alternative='two-sided')
            for d in d_values]
powers

[0.07895538929132649,
 0.16948092445426557,
 0.32174707632764465,
 0.5140816347005528,
 0.703333297940797,
 0.8491491055035358,
 0.9371867266873937,
 0.9788403128661938,
 0.9942830780694515,
 0.9987688466795083]

Looking at this chart, we see that power really starts to drop off
around *d* = .20, hitting 51%. You discuss this with your team; they
conclude they are ok with a 50% chance of declaring "no difference" if
effect is *that* small. The study is run with 400 people and an
informative result is produced.

Epilogue
========

At the very end, you hear that the original proposal had been to collect
40 responses, 20 from each site. You smile to yourself, considering how
your power analysis likely saved the project. You run a power analysis
just to see how bad the situation would be:

In [6]:
tt_ind_solve_power(effect_size=None, nobs1 = 20, alpha=0.05, power=0.8, ratio=1, alternative='two-sided')

0.9091290568795592

You see here that the smallest effect size for which you would have good
power is well into the 'large' range. What if the effects were small?
What are the odds the study would even be able to pick them up? You
consider the scenario of *d* = .25:

In [7]:
tt_ind_solve_power(effect_size=.25, nobs1 = 20, alpha=0.05, power=None, ratio=1, alternative='two-sided')

0.12033536754525204

This study would have 12% power. Yikes. It's a good thing you performed
a power analysis.